### PR TITLE
Issue 2525: Fix lint errors

### LINKED ISSUE #2525: Linter was not running the whole time
I was looking at a PR and noticed an unused variable was not detected by `npm run lint`
**Shocking discovery 🙃**

Our linting script `eslint --config .eslintrc.js` does not provide a glob pattern to ESLint and no files are linted.

Developers only notice problems found by ESLint if they have installed the VSCode ESLint extension because the extension puts squiggly underlines on them. Otherwise, running `npm run lint` doesn't show any linting problems.

Plus, we use some Eslint plugins that are not listed in `package.json`. We don't notice this because npm run lint was not running the whole time.

This can explain [the comment](https://github.com/Seneca-CDOT/telescope/pull/2420#pullrequestreview-793706259) by @humphd why the new dashboard passed all eslint tests.

### PR DESCRIPTION
**Pull request by DukeManh**
## Issue This PR Addresses
Fixes #2525 

This fixes some linting errors that were left out due to Eslint not being provided with a target files pattern.

These are some common errors in the code.
1. Using \<a target=“_blank”\> without rel=“referrer” is a [security risk](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types/noreferrer)
2. Empty components are self-closing
3. [Prevent invalid characters from appearing in markup](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-unescaped-entities.md#prevent-invalid-characters-from-appearing-in-markup-reactno-unescaped-entities)
4. [No useless escape](https://eslint.org/docs/rules/no-useless-escape)
5. [Disallow continue statements](https://eslint.org/docs/rules/no-continue)
6. [Avoid for..of loop in favor of array iterator](https://medium.com/@paul.beynon/thanks-for-taking-the-time-to-write-the-article-i-enjoyed-it-db916026647)
7. [Consistent return, either always or never](https://eslint.org/docs/rules/consistent-return#require-return-statements-to-either-always-or-never-specify-values-consistent-return)
8. [No restricted globals](https://eslint.org/docs/rules/no-restricted-globals#disallow-specific-global-variables-no-restricted-globals)
9. [No use before define](https://eslint.org/docs/rules/no-use-before-define#disallow-early-use-no-use-before-define)
10. No multiple chained assignment
11. No unused variables
12. [React Hooks must not be called conditionally](https://reactjs.org/docs/hooks-rules.html)

## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [ ] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Checklist

- [ ] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -1,3 +1,12 @@
+// React is installed in src/web but Eslint is unable to detect it
+const { react } = require('./src/web/package.json').dependencies;
+
+// Extract react version manually to pass it to Eslint
+// Semver Regex: https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
+const reactVersion = react.match(
+  /(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$/
```
Comment (humphd) [inline .eslintrc.js:7]: This is horrible.  Can we just not check the React version?

File: package.json [hunk 2]
```diff
@@ -12,15 +12,17 @@
     "install:users-service": "cd src/api/users && npm install",
     "install:posts-service": "cd src/api/posts && npm install",
     "install:search-service": "cd src/api/search && npm install",
+    "install:planet-service": "cd src/api/planet && npm install",
+    "install:status-service": "cd src/api/status && npm install",
```
Comment (humphd) [inline package.json:16]: Nice catch

File: package.json [hunk 3]
```diff
@@ -12,15 +12,17 @@
...
     "install:feed-discovery-service": "cd src/api/feed-discovery && npm install",
     "install:parser-service": "cd src/api/parser && npm install",
     "install:autodeployment": "cd tools/autodeployment && npm install",
     "install:next": "cd src/web && npm install",
     "build": "npm run build --prefix src/web --",
     "dev": "npm run dev --prefix src/web --",
     "develop": "npm run dev",
-    "eslint": "eslint --config .eslintrc.js",
-    "eslint-fix": "eslint --config .eslintrc.js --fix",
+    "eslint": "eslint --config .eslintrc.js \"**/*.{jsx,tsx,ts,js}\"",
```
Comment (humphd) [inline package.json:24]: 🤓 How did we miss this!?

File: src/api/users/test/e2e/users.test.js [hunk 4]
```diff
@@ -73,7 +73,6 @@ describe('GET REQUESTS', () => {
   it('should 400, rejected by negative per_page', async () => {
     // start_after does not matter, middleware should reject regardless
     const response = await getUsersPaginated('?per_page=-5');
-    const users = await response.json();
```
Comment (humphd) [inline src/api/users/test/e2e/users.test.js:76]: Interesting

**Review by humphd — CHANGES_REQUESTED** [review]
Looks great, a couple things.

File: .eslintrc.js (hunk 1 above)
Comment (DukeManh) [reply .eslintrc.js:7]: I can just hard-code React v17.0.

Comment (humphd) [reply .eslintrc.js:7]: Let's do that and fix it when we upgrade React later

File: src/web/src/components/NavBar/index.tsx [hunk 5]
```diff
@@ -150,9 +150,9 @@ export default function NavBar({ disabled }: NavBarProps) {
               <DynamicThemeToggleButton />
               {user && (
                 <ButtonTooltip title="Logout" arrow placement="top" TransitionComponent={Zoom}>
-                  <div className={classes.avatar} onClick={() => logout()}>
+                  <button type="button" className={classes.avatar} onClick={() => logout()}>
```
Comment (Andrewnt219) [inline src/web/src/components/NavBar/index.tsx:153]: Nice catch, `div onClick` is a bad a11y design.

File: src/api/users/test/e2e/users.test.js (hunk 4 above)
Comment (DukeManh) [reply src/api/users/test/e2e/users.test.js:76]: It's an unused variable.

File: tools/autodeployment/builds.js [hunk 6]
```diff
@@ -112,9 +112,9 @@ module.exports.buildLogHandler = function handleLog(req, res, next) {
 
   res.writeHead(200, { 'Content-Type': 'text/plain' });
 
-  const onData = (data) => res.write(data);
-  const onError = () => end('Error, end of log.');
-  const onEnd = () => end('Build Complete.');
+  let onData;
+  let onError;
+  let onEnd;
```
Comment (Andrewnt219) [inline tools/autodeployment/builds.js:117]: Isn't the old code more precise?

**Review by Andrewnt219 — CHANGES_REQUESTED** [review]
Good job, that's a lot of work done, One question: is it an eslint rule to have a return In all functions?

File: src/api/status/public/js/build-log/check-for-build.js [hunk 7]
```diff
@@ -1,6 +1,9 @@
-import { checkBuildStatus, getBuildLog } from './api.js';
-import terminal from './terminal.js';
-let build, reader;
+/* eslint-disable consistent-return */
+import { checkBuildStatus, getBuildLog } from './api';
+import terminal from './terminal';
```
[the remark this answers is not in the document]
Comment (DukeManh) [reply src/api/status/public/js/build-log/check-for-build.js:3]: Eslint thought this is running on Node. Node has an automatic extension resolver. I'll see if can exclude this folder.

File: tools/autodeployment/builds.js (hunk 6 above)
Comment (DukeManh) [reply tools/autodeployment/builds.js:117]: The `end` function is used before its definition.  It works but ESLint recommends not to do that. 
https://eslint.org/docs/rules/no-use-before-define

File: src/api/status/public/js/pages/build.js [hunk 8]
```diff
@@ -1,4 +1,4 @@
-import { checkForBuild } from '../build-log/check-for-build.js';
+import { checkForBuild } from '../build-log/check-for-build';
```
[the remark this answers is not in the document]
Comment (DukeManh) [reply src/api/status/public/js/pages/build.js:1]: https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/prefer-default-export.md
It's just consistent style enforcement.

**Comment by DukeManh** [comment]
@Andrewnt219
> One question: is it an eslint rule to have a return In all functions?

It's this rule: https://eslint.org/docs/rules/consistent-return#require-return-statements-to-either-always-or-never-specify-values-consistent-return

So basically, we're required to either have no return or return something in every branch of the function.

File: src/api/status/public/js/pages/build.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (Andrewnt219) [reply src/api/status/public/js/pages/build.js:1]: Ok, then you need to import default here too.

File: tools/autodeployment/builds.js (hunk 6 above)
Comment (Andrewnt219) [reply tools/autodeployment/builds.js:117]: oh, that's right, I missed the `end` declaration.

**Comment by Andrewnt219** [comment]
> So basically, we're required to either have no return or return something in every branch of the function.

I see. Because the PR changes functions that return void to return a value anyway. It isn't a problem though, maybe it's better to be explicit like this.

File: src/api/status/public/js/build-log/check-for-build.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (DukeManh) [reply src/api/status/public/js/build-log/check-for-build.js:3]: It should always require import extension in `status/public` now

**Review by Andrewnt219 — APPROVED** [review]
LGTM.

**Review by humphd — APPROVED** [review]
Epic.  Thank you for a) noticing this; b) fixing it.

**Pull request merged**